MSC2398: proposal to allow mxc:// in the "a" tag within messages (#2398)
matrix-org/matrix-doc: proposals/2398-allow-mxc-within-message-links.md

> # MSC2398: Proposal to allow mxc:// in the `a` tag within messages

Reviewer: The media ID would be attributed to a single event somewhere, and clients would be translating `mxc://` to `https://` for the user. This then behaves no differently than if a user were to get linked to the download endpoint directly - authentication would be required, and if that user can't see the original event then the link is broken for that user.

> # MSC2398: Proposal to allow mxc:// in the `a` tag within messages

Reviewer 2: > This then behaves no differently than if a user were to get linked to the download endpoint directly - authentication would be required, and if that user can't see the original event then the link is broken for that user.

But getting linked directly to the download endpoint won't work, as there is no way for a user to authenticate AIUI? I thought for media web clients would potentially need to download the media in service workers or whatever to make the authentication work, in a similar manner as today to make encrypted attachments work. Is that method viable here?

> # MSC2398: Proposal to allow mxc:// in the `a` tag within messages

Reviewer: A service worker would be required, which should be workable. The client can essentially iframe it and append the appropriate headers. This then ties into Matthew's comment about security considerations: https://github.com/matrix-org/matrix-spec-proposals/pull/2398#discussion_r1450850788

I'm tempted to suggest a blocking concern on this as it feels valuable to test/implement, particularly while MSC3916 and friends are being experimented with. It's akin to custom emoji/stickers where I think we want to avoid making the problem worse short-term if we can avoid it.

Currently message events (m.room.message) can contain formatted data

> in its formatted_body field, if the format of the message is

Reviewer 3: ```suggestion
in their formatted_body field, if the format of the message is
```

> org.matrix.custom.html. In this subset of HTML `img` tags are permitted

Reviewer 3: ```suggestion
org.matrix.custom.html. In this subset of HTML, `img` tags are permitted
```

if they refer to media repository contents with the mxc schema and

> clients are able to resolve these URIs to (ie.) HTTPs to retrieve the

Reviewer 3: ```suggestion
clients are able to resolve these URIs to (eg.) HTTP(s) URLs to retrieve the
```

Sorry if these feel nitpicky, I'm just having to re-read a couple of times to work out what the MSC is saying so attempting to clarify.

actual contents.

However, the spec currently forbids `a` tags from receiving the same
treatment. These links could be useful for example for systems that
upload PDF files to the media storage and then produce messages
linking to such files.

## Proposal

> The spec for `a` tags says:

Reviewer: ```suggestion
The spec for `a` tags [says](https://spec.matrix.org/v1.8/client-server-api/#mroommessage-msgtypes):
```

-`a`:
  `name`, `target`, `href` (provided the value is not relative and has a scheme
  matching one of: `https`, `http`, `ftp`, `mailto`, `magnet`)

Clients would be able to handle the mxc schema here in the exact same
way they handle `img` tags. Thus the proposal is to modify that fragment
to say:

-`a`:
  `name`, `target`, `href` (provided the value is not relative and has a scheme
  matching one of: `https`, `http`, `ftp`, `mailto`, `magnet`, or `mxc`

>   for `Matrix Content (MXC) URI`)

Reviewer 3: Much as it's obvious, presumably the spec would also need to explicitly say that the client must transform the mxc link into an appropriate link the browser will understand (ie. what you talk about above with img tags).

Reviewer: This feels like something which is best served by a note rather than requirement. Web clients will probably need to convert it, but other clients might be able to handle a custom scheme for network requests. 

I'd suggest that this *not* be a blocking concern for FCP, as the spec text can provide guidance on how to handle it.

Reviewer 4: I would really prefer the MSC say explicitly that clients should manage this scheme in one way or another - ultimately a custom handler will have to translate the `mxc` URL into an HTTP(S) request anyway.

Reviewer 4: Oh wait but isn't "Clients would be able to handle the mxc schema here in the exact same way they handle `img` tags" above exactly this description, actually? I personally think that is enough.

Reviewer 5: What do you think @dbkr?

Reviewer 3: I guess it just seemed odd to me that an explicit change to the spec text was given but then additional behaviour specified in the MSC. I suppose it's all there though so it doesn't really matter.

## Potential issues ##

As far as I see, this is quite a benign change.

The only problem I see (in addition to possible problems already
present in `img` tags) is that current clients may not be able to handle
mxc tags in this position; thus such clients would not be able to
retrieve resources linked this way. Some clients may already
accidentally support this, though, due to consistent handling of URIs.

> ## Security considerations ##

Reviewer 6: It's probably worth spelling out explicitly here that the security risks of linking directly to content in the repository is not just HTML, but any format which can execute code and potentially XSS needs to be mitigated by a valid CSP - e.g. JS, PDF, SWF, SVG.  This isn't anything new, but it's still a security consideration.  I can't think of a world where a HTTPS url derived from an MXC URL wouldn't be subject to CSP when a plain HTTPS url to the same content would, but still worth spelling it out at least in the MSC, if not the final spec.
